### PR TITLE
ci: use workspace inheritance to enforce lints in all crates

### LINKED ISSUE #4484: replace `custom-clippy` alias with workspace-wide `[lints]` configuration
Currently, we have an alias for standardizing our lint configuration:

https://github.com/libp2p/rust-libp2p/blob/240019f8099142354173d252bbb9d6fffd40be0f/.cargo/config.toml#L3

There is a new cargo feature on the way that allows to configure lints in the `Cargo.toml`. Together with workspace-inheritance,  we should be able to move our lint configuration away from the alias to a configuration that is only specified once in our root `Cargo.toml`.

Here is the tracking issue: https://github.com/rust-lang/cargo/issues/12115

This should stable from `nightly-2023-09-10` onwards. We should be able to start experimenting with this by adding the relevant nightly toolchain here: https://github.com/libp2p/rust-libp2p/blob/240019f8099142354173d252bbb9d6fffd40be0f/.github/workflows/ci.yml#L202-L205 and moving the configuration from `custom-clippy` over to `Cargo.toml`.

Contributions welcome!

### PR DESCRIPTION
**Pull request by binsta**
## Description

Starting with nightly-2023-09-10, the `[lints]` section in `Cargo.toml` files is stable. Together with workspace inheritance, this can be used to declare all lints we want to enforce in a single place.

Resolves: #4484.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by binsta** [comment]
@thomaseizinger Please take a look. If any changes are needed, please let me know. if the `[lints]` tag is in a stable release it will cool.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -201,7 +201,8 @@ jobs:
       matrix:
         rust-version: [
           1.72.0, # current stable
-          beta
+          beta,
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml]: Can you comment out beta and stable here, delete the alias and just run `cargo clippy` in this job?

File: Cargo.toml [hunk 2]
```diff
@@ -128,3 +128,10 @@ multihash = "0.19.1"
 # we import via `rust-multiaddr`.
 # This is expected to stay here until we move `libp2p-identity` to a separate repository which makes the dependency relationship more obvious.
 libp2p-identity = { path = "identity" }
+
+# lint 
+[workspace.lints.clippy]
```
Comment (thomaseizinger) [inline Cargo.toml]: Does it work like this or do we need to use workspace inheritance in all crates to actually apply this?

**Comment by thomaseizinger** [comment]
> @thomaseizinger Please take a look. If any changes are needed, please let me know. if the `[lints]` tag is in a stable release it will cool.

Thanks for getting this started :)

File: Cargo.toml (hunk 2 above)
Comment (binsta) [reply Cargo.toml]: yes, lot of cargo.toml to cover 
```
[lints]
workspace = true
```
is It fine or any suggestion

File: .github/workflows/ci.yml (hunk 1 above)
Comment (binsta) [reply .github/workflows/ci.yml]: @thomaseizinger Just now commit the change. Please check

**Comment by binsta** [comment]
@thomaseizinger  I have a quick question on `unreachable_pub`. Running on cargo +nightly clippy 
```
warning[E0602]: unknown lint: `clippy::unreachable_pub
```
Please check [clippy::useless_attribute](https://rust-lang.github.io/rust-clippy/master/index.html#/useless_attribute)

**Comment by thomaseizinger** [comment]
> @thomaseizinger I have a quick question on `unreachable_pub`. Running on cargo +nightly clippy
> 
> ```
> warning[E0602]: unknown lint: `clippy::unreachable_pub
> ```
> 
> Please check [clippy::useless_attribute](https://rust-lang.github.io/rust-clippy/master/index.html#/useless_attribute)

That is because it is a `rustc` lint, not a clippy lint! I think you'll have to move it out of the clippy group in `Cargo.toml` :)

File: Cargo.toml (hunk 2 above)
Comment (thomaseizinger) [reply Cargo.toml]: That looks like what we need, yes!

File: core/Cargo.toml [hunk 3]
```diff
@@ -50,3 +50,7 @@ serde = ["multihash/serde-codec", "dep:serde", "libp2p-identity/serde"]
 all-features = true
 rustdoc-args = ["--cfg", "docsrs"]
 rustc-args = ["--cfg", "docsrs"]
+
+# lint
```
Comment (thomaseizinger) [inline core/Cargo.toml]:
```suggestion
```

Don't think we need this comment :)

File: core/Cargo.toml [hunk 4]
```diff
@@ -50,3 +50,7 @@ serde = ["multihash/serde-codec", "dep:serde", "libp2p-identity/serde"]
...
+[lints]
+workspace = true
```
Comment (thomaseizinger) [inline core/Cargo.toml]: Can you ensure there are newlines at the end of files? See the red marker here in the diff. You can probably configure your editor to automatically add them.

File: core/Cargo.toml (hunk 3 above)
Comment (binsta) [reply core/Cargo.toml]: Okay I have removed the comment

File: core/Cargo.toml (hunk 4 above)
Comment (binsta) [reply core/Cargo.toml]: Sure, Just now made a commit, Please review. Sorry for the delay.

File: Cargo.toml [hunk 5]
```diff
@@ -128,3 +128,10 @@ multihash = "0.19.1"
 # we import via `rust-multiaddr`.
 # This is expected to stay here until we move `libp2p-identity` to a separate repository which makes the dependency relationship more obvious.
 libp2p-identity = { path = "identity" }
+
+# lint
```
Comment (thomaseizinger) [inline Cargo.toml]:
```suggestion
```

**Comment by thomaseizinger** [comment]
@binsta Once you've resolved the merge conflicts, CI can run again :)

**Comment by thomaseizinger** [comment]
> @binsta Once you've resolved the merge conflicts, CI can run again :)

Btw, merging master is fine, we don't really do rebases here! See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#we-squash-merge-pull-requests.

**Comment by binsta** [comment]
@thomaseizinger  Please review every conflict that has been resolved.

**Comment by binsta** [comment]
> > @binsta Once you've resolved the merge conflicts, CI can run again :)
> 
> Btw, merging master is fine, we don't really do rebases here! See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#we-squash-merge-pull-requests.

Thank you, Hereafter ill follow the same, methodology

**Comment by thomaseizinger** [comment]
> @thomaseizinger Please review every conflict that has been resolved.

Thanks! They weren't quite resolved but I went ahead and fixed them for you :)

File: examples/browser-webrtc/Cargo.toml [hunk 6]
```diff
@@ -21,9 +21,18 @@ rand = "0.8"
...
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
 axum = "0.6.19"
-libp2p = { path = "../../libp2p", features = ["ed25519", "macros", "ping", "wasm-bindgen", "tokio"] }
+libp2p = { path = "../../libp2p", features = [
+    "ed25519",
+    "macros",
+    "ping",
+    "wasm-bindgen",
+    "tokio",
+] }
```
Comment (thomaseizinger) [inline examples/browser-webrtc/Cargo.toml]: Please avoid these re-formattings, we don't currently enforce a formatting in CI for toml files so I'd like to keep the noise to a minimum! If you are interested, you could look into a solution to https://github.com/libp2p/rust-libp2p/issues/4043 :)

**Review by thomaseizinger — COMMENTED** [review]
CI for clippy is passing which is nice! See https://github.com/libp2p/rust-libp2p/actions/runs/6386769096/job/17333995887?pr=4575.

I think we can merge this as is and simply swap out the clippy jobs again once this hits stable!

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -218,7 +219,7 @@ jobs:
           save-if: ${{ github.ref == 'refs/heads/master' }}
 
       - name: Run cargo clippy
-        run: cargo custom-clippy # cargo alias to allow reuse of config locally
+        run: cargo clippy # cargo alias to allow reuse of config locally
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml]:
```suggestion
        run: cargo clippy
```

Can you also delete the cargo alias in `.cargo/`? I think we may be able to get rid of the entire file!

Comment (binsta) [reply .github/workflows/ci.yml]: sure

File: examples/browser-webrtc/Cargo.toml (hunk 6 above)
Comment (binsta) [reply examples/browser-webrtc/Cargo.toml]: okay. I'll take a look #4043, changes made as per requested.

File: Cargo.toml (hunk 5 above)
Comment (binsta) [reply Cargo.toml]: Removed the comment in Cargo.toml

**Comment by binsta** [comment]
@thomaseizinger  Please take a look

File: .github/workflows/ci.yml (hunk 7 above)
Comment (thomaseizinger) [inline .github/workflows/ci.yml]:
```suggestion
        run: cargo clippy
```

File: protocols/autonat/Cargo.toml [hunk 8]
```diff
@@ -4,7 +4,10 @@ edition = "2021"
 rust-version = { workspace = true }
 description = "NAT and firewall detection for libp2p"
 version = "0.11.0"
-authors = ["David Craven <david@craven.ch>", "Elena Frank <elena.frank@protonmail.com>"]
+authors = [
+    "David Craven <david@craven.ch>",
+    "Elena Frank <elena.frank@protonmail.com>",
+]
```
Comment (thomaseizinger) [inline protocols/autonat/Cargo.toml]:
```suggestion
authors = ["David Craven <david@craven.ch>", "Elena Frank <elena.frank@protonmail.com>"]
```

File: swarm-derive/Cargo.toml [hunk 9]
```diff
@@ -16,7 +16,13 @@ proc-macro = true
...
 heck = "0.4"
 quote = "1.0"
-syn = { version = "2.0.37", default-features = false, features = ["clone-impls", "derive", "parsing", "printing", "proc-macro"] }
+syn = { version = "2.0.37", default-features = false, features = [
+    "clone-impls",
+    "derive",
+    "parsing",
+    "printing",
+    "proc-macro",
+] }
```
Comment (thomaseizinger) [inline swarm-derive/Cargo.toml]:
```suggestion
syn = { version = "2.0.37", default-features = false, features = ["clone-impls", "derive", "parsing", "printing", "proc-macro"] }
```

File: swarm/Cargo.toml [hunk 10]
```diff
@@ -24,7 +24,9 @@ rand = "0.8"
 smallvec = "1.11.1"
 void = "1"
 wasm-bindgen-futures = { version = "0.4.37", optional = true }
-getrandom = { version = "0.2.9", features = ["js"], optional = true } # Explicit dependency to be used in `wasm-bindgen` feature
+getrandom = { version = "0.2.9", features = [
+    "js",
+], optional = true } # Explicit dependency to be used in `wasm-bindgen` feature
```
Comment (thomaseizinger) [inline swarm/Cargo.toml]:
```suggestion
getrandom = { version = "0.2.9", features = ["js"], optional = true } # Explicit dependency to be used in `wasm-bindgen` feature
```

File: transports/pnet/Cargo.toml [hunk 11]
```diff
@@ -20,7 +20,7 @@ pin-project = "1.1.3"
 
 [dev-dependencies]
 libp2p-core = { workspace = true }
-libp2p-identity = { workspace = true, features = ["ed25519", "rsa", "ecdsa", "secp256k1"] }
+libp2p-identity = { workspace = true, features = [ "ed25519", "rsa", "ecdsa","secp256k1"] }
```
Comment (thomaseizinger) [inline transports/pnet/Cargo.toml]:
```suggestion
libp2p-identity = { workspace = true, features = ["ed25519", "rsa", "ecdsa","secp256k1"] }
```

File: transports/quic/Cargo.toml [hunk 12]
```diff
@@ -19,11 +19,11 @@ libp2p-tls = { workspace = true }
 libp2p-identity = { workspace = true }
 log = "0.4"
 parking_lot = "0.12.0"
-quinn = { version = "0.10.2", default-features = false, features = ["tls-rustls", "futures-io"] }
+quinn = { version = "0.10.2", default-features = false, features = [ "tls-rustls", "futures-io"] }
```
Comment (thomaseizinger) [inline transports/quic/Cargo.toml]:
```suggestion
quinn = { version = "0.10.2", default-features = false, features = ["tls-rustls", "futures-io"] }
```

File: transports/quic/Cargo.toml [hunk 13]
```diff
@@ -19,11 +19,11 @@ libp2p-tls = { workspace = true }
...
 rand = "0.8.5"
 rustls = { version = "0.21.7", default-features = false }
 thiserror = "1.0.48"
-tokio = { version = "1.32.0", default-features = false, features = ["net", "rt", "time"], optional = true }
+tokio = { version = "1.32.0", default-features = false, features = [ "net", "rt", "time"], optional = true }
```
Comment (thomaseizinger) [inline transports/quic/Cargo.toml]:
```suggestion
tokio = { version = "1.32.0", default-features = false, features = ["net", "rt", "time"], optional = true }
```

File: transports/tcp/Cargo.toml [hunk 14]
```diff
@@ -20,7 +20,7 @@ libp2p-core = { workspace = true }
 libp2p-identity = { workspace = true }
 log = "0.4.20"
 socket2 = { version = "0.5.4", features = ["all"] }
-tokio = { version = "1.32.0", default-features = false, features = ["net"], optional = true }
+tokio = { version = "1.32.0", default-features = false, features = ["net",], optional = true }
```
Comment (thomaseizinger) [inline transports/tcp/Cargo.toml]:
```suggestion
tokio = { version = "1.32.0", default-features = false, features = ["net"], optional = true }
```

File: transports/tls/Cargo.toml [hunk 15]
```diff
@@ -30,7 +30,7 @@ features = ["dangerous_configuration"] # Must enable this to allow for custom ve
 hex = "0.4.3"
 hex-literal = "0.4.1"
 libp2p-core = { workspace = true }
-libp2p-identity = { workspace = true, features = ["ed25519", "rsa", "secp256k1", "ecdsa"] }
+libp2p-identity = { workspace = true, features = [ "ed25519", "rsa", "secp256k1", "ecdsa"] }
```
Comment (thomaseizinger) [inline transports/tls/Cargo.toml]:
```suggestion
libp2p-identity = { workspace = true, features = ["ed25519", "rsa", "secp256k1", "ecdsa"] }
```

File: transports/webrtc-websys/Cargo.toml [hunk 16]
```diff
@@ -28,9 +28,12 @@ serde = { version = "1.0", features = ["derive"] }
 thiserror = "1"
 wasm-bindgen = { version = "0.2.87" }
 wasm-bindgen-futures = { version = "0.4.37" }
-web-sys = { version = "0.3.64", features = ["Document", "Location", "MessageEvent", "Navigator", "RtcCertificate", "RtcConfiguration", "RtcDataChannel", "RtcDataChannelEvent", "RtcDataChannelInit", "RtcDataChannelState", "RtcDataChannelType", "RtcPeerConnection", "RtcSdpType", "RtcSessionDescription", "RtcSessionDescriptionInit", "Window"] }
+web-sys = { version = "0.3.64", features = [ "Document", "Location", "MessageEvent", "Navigator", "RtcCertificate", "RtcConfiguration", "RtcDataChannel", "RtcDataChannelEvent", "RtcDataChannelInit", "RtcDataChannelState", "RtcDataChannelType", "RtcPeerConnection", "RtcSdpType", "RtcSessionDescription", "RtcSessionDescriptionInit", "Window"] }
```
Comment (thomaseizinger) [inline transports/webrtc-websys/Cargo.toml]:
```suggestion
web-sys = { version = "0.3.64", features = ["Document", "Location", "MessageEvent", "Navigator", "RtcCertificate", "RtcConfiguration", "RtcDataChannel", "RtcDataChannelEvent", "RtcDataChannelInit", "RtcDataChannelState", "RtcDataChannelType", "RtcPeerConnection", "RtcSdpType", "RtcSessionDescription", "RtcSessionDescriptionInit", "Window"] }
```

**Review by thomaseizinger — DISMISSED** [review]
Thanks! This is good to go from my end :)

**Comment by thomaseizinger** [comment]
@libp2p/rust-libp2p-maintainers: Starting with this PR, you'll have to use nightly Rust for linting locally for a limited time. It is generally a good idea to use nightly Rust for local development because it helps detecting compiler bugs early. Our CI enforces that things work with stable anyway.

This feature is due to be stabilized in 1.74 I think hence I think it makes sense to already start using this. Once it is actually stable, we can switch CI back to running on the current stable and beta.

If there are no concerns, I'll merge this in the upcoming days.

**Comment by thomaseizinger** [comment]
> @libp2p/rust-libp2p-maintainers: Starting with this PR, you'll have to use nightly Rust for linting locally for a limited time. It is generally a good idea to use nightly Rust for local development because it helps detecting compiler bugs early. Our CI enforces that things work with stable anyway.
> 
> This feature is due to be stabilized in 1.74 I think hence I think it makes sense to already start using this. Once it is actually stable, we can switch CI back to running on the current stable and beta.
> 
> If there are no concerns, I'll merge this in the upcoming days.

Most notably, nothing changes for users that are currently using stable rust and run `cargo clippy`. They'll still see a bunch of warning that are only ignored via our `custom-clippy` alias. To transition those, we could retain the alias for now but change it to run on a specific Rust toolchain, i.e. the nightly version that stabilizes the `[lints]` feature.

Or we just retain the alias altogether until `[lints]` stabilizes completely. We just need to manually keep it in sync until then but that shouldn't be too difficult given how rarely it changes.

**Review by mxinden — COMMENTED** [review]
As discussed in our 1on1, feel free to go ahead here @thomaseizinger.

**Review by thomaseizinger — DISMISSED** [review]

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**